content/recipes/async-local-storage.md

```diff
@@ -0,0 +1,225 @@
+### Async Local Storage
+
+`AsyncLocalStorage` is a [Node.js API](https://nodejs.org/api/async_context.html#async_context_class_asynclocalstorage) (based on the `async_hooks` API) that provides an alternative way of propagating local state through the application without the need to explicitly pass it as a function parameter. It is similar to a thread-local storage in other languages.
+
+The main idea of Async Local Storage is that we can _wrap_ some function call with the `AsyncLocalStorage#run` call. All code that is invoked within the wrapped call gets access to the same `store`, which will be unique to each call chain.
+
+In the context of NestJS, that means if we can find a place within the request's lifecycle where we can wrap the rest of the request's code, we will be able to access and modify state visible only to that request, effectively eliminating the need for REQUEST scoped providers and their limitations.
```

Comment:
REQUEST-scoped providers have some other great use cases, like constructing providers tied to specific requests so we might need to mention here that these two approaches are not really interchangeable.

Reply:
Constructing providers per request is also possible with the Proxy Providers feature of `nestjs-cls`, but yeah, I'll try to reword it, so it's not so harsh on REQUEST-scoped providers.

Reply:
Sure, but REQUEST-scoped providers have limitations and simply cannot be used in certain cases, and that's the whole point of using AsyncLocalStorage, which doesn't require consuming providers to also become request-scoped.

With proxy providers, I can create the same "request" scoped provider for a CRON job as well as for a HTTP request and consume it in my services using the same API (any my services don't need to be aware of the notion of a request, hence separating concerns). This is not re-inventing a wheel per-se, it's just offering a different set of wheels for a similar functionality.

Don't get me wrong, I'm not trying to undermine the features of the framework, but I'll make a bold statement that ALS can completely replace REQUEST-scoped providers. And when (if?) ALS matures enough to be entirely stable and introduce neglibigle runtime overhead, we might even go as far as to reimplement actual REQUEST-scoped providers on top of it in NestJS itself.

Reply:
> With proxy providers, I can create the same "request" scoped provider for a CRON job as well as for a HTTP request or Kafka message and consume it in my services using the same API (any my services don't need to be aware of the notion of doferent transports, hence separating concerns). This is not re-inventing a wheel per-se, it's just offering a different set of wheels for a similar functionality.

I don't see how this feature differs from the built-in request-scoped providers. This should be already feasible unless I misunderstood what you're trying to accomplish.

> But I'll make a bold statement that ALS can completely replace REQUEST-scoped providers.

I'm not saying that it _can't completely replace them_. I'm just saying that _it shouldn't_ because it _makes no sense_.

Reply:
Also, just to make sure we're on the same page. I'm not saying that using ALS is wrong by any means. ALS is great and it has some nice use cases. What I'm trying to emphasize is - using ALS to construct request-scoped providers doesn't seem to be the cleanest and optimal solution when you can already accomplish what you need by utilizing custom DI sub-trees (if you need dynamic instantiation, transport strategy agnostic) or with simple request-scoped providers (if the static hierarchy is fine) 

ALS and request-scoped providers are not interchangeable. Ideally, we could (and would want, whenever ALS becomes stable and mature enough) leverage both depending on what we want to accomplish (possibly even at the same time).

Reply:
> I don't see how this feature differs from the built-in request-scoped providers. This should be already feasible unless I misunderstood what you're trying to accomplish.

Imagine I have a multi tenant app, which has multiple entry points, but they all eventually use the same set of shared services. There are the following requirements:
1) limit access only to records based on the `tenantId` of the request,
2) restrict certain functionality based on the `userRole of the user making that request`

I could pass these variables down to every method or event, but you understand that it would be an absolute nightmare to maintain and re-factor.

So, using `nestjs-cls`, I create a Proxy Provider called "RequestContext" that has 2 properties - `userRole` and `tenantId`. 

In my app, I have HTTP endpoints, Kafka consumers, CRON controllers. Each of which has a different way of extracting the context from the message and populating the ALS. I can feed the ALS (in three different ways) with it and construct the `RequestContext` object based on the values in ALS. The rest of the app only depends on the `RequestContext` provider.

There's no REQUEST in Kafka, there's only CONTEXT - so I don't see how I could create a shared provider based on both.

Additionally, last I checked, there was no way to use services that depend on the REQUEST in a CRON handler.

So, how would you do something like this using built-in REQUEST-scoped providers?

(Sorry, I feel like I'm diverging significantly from the original topic 😅, but I'd love to have a discussion about application design sometime)

Reply:
> There's no REQUEST in Kafka, there's only CONTEXT - so I don't see how I could create a shared provider based on both.

`CONTEXT` is just an alias for `REQUEST`. You could create a `RequestContext` factory provider that injects `REQUEST` and extracts "the context data" based on the transport layer.

> Additionally, last I checked, there was no way to use services that depend on the REQUEST in a CRON handler.

AFAIR it's feasible with Bull crons, I can't recall if we support it out-of-the-box with nestjs/schedule. Still, you can dynamically instantiate request-scoped providers from statically-scoped providers using custom DI sub-trees (and registering a custom, platform-agnostic RequestContext object for that subtree).